processor/src/main/java/com/linecorp/decaton/processor/runtime/internal/ProcessPipeline.java

```diff
@@ -76,7 +77,7 @@ public void scheduleThenProcess(TaskRequest request) throws InterruptedException
         }
 
         scheduler.schedule(extracted.metadata());
-        if (terminated) {
+        if (terminated && 0L == scope.props().get(ProcessorProperties.CONFIG_CLOSE_TIMEOUT_MS).value()) {
```

Comment:
As in the purpose of this PR is to avoid completing shutdown of subscription while some tasks are still in progress (i.e, tasks already passed to `DecatonProcessor#process` but haven't completed yet), shouldn't we stop feeding more tasks in to processor at the point the termination flag set on?
Hm, but that'll make it unable to work just calling `waitForRemainingTasks...` at shutdown because it will not complete, rather we have to count the tasks we've passed to `DecatonProcessor` yet hasn't complete.
But since user initiated the shutdown sequence, no one's happy with the behavior that keeps feeding tasks in queue despite termination has begun which just extend the duration required to finish shutdown?

Reply:
I've reconsidered this, I think I should withdraw this > shouldn't we stop feeding more tasks in to processor at the point the termination flag set on?

The behavior (wait all "fetched" tasks until all of them finished processing)  is what we're already doing at rebalance, which also has timeout, makes us to assume that users set `decaton.max.pending.records` to appropriate value to make sure the processing completes before `decaton.rebalance.timeout.ms` elapses. So it might be okay to assume the same premise on shutdown time too.

Reply:
Btw, this change still has a different problem.

`scheduler.schedule()` may returns earlier than the originally planned time when it's `close()` method is called. So when 1. `scheduler.schedule()` returns and 2. terminated flag is on, we should not assume the task has reached the time ready to start processing, so we need more phased shutdown sequence to obtain the desired behavior.

1. `ProcessorSubscription.initateShutdown()` called
2. Stop feeding more requests from the consumer (exit pool loop or pause all partitions)

Reply:
> The behavior (wait all "fetched" tasks until all of them finished processing) is what we're already doing at rebalance, which also has timeout, makes us to assume that users set decaton.max.pending.records to appropriate value to make sure the processing completes before decaton.rebalance.timeout.ms elapses. So it might be okay to assume the same premise on shutdown time too.

Right. I think for a "graceful" shutdown we should try to process all fetched tasks - in particular, we might not be able to commit offsets for completed tasks if we don't process other tasks that were fetched at the same time (because of out-of-order processing).

> So we don't actually need condition added I think.

Ah, you're right. I'll remove it.

I'm a bit concerned that this means that one scheduled task will block us from committing completion of any tasks that arrived later. But I guess that's already the behaviour on rebalance.